sktime/base/_base.py

```diff
@@ -111,9 +114,126 @@ def clone(self):
         A clone is a different object without shared references, in post-init state.
         This function is equivalent to returning sklearn.clone of self.
         Equal in value to `type(self)(**self.get_params(deep=False))`.
+
+        Returns
+        -------
+        instance of type(self), clone of self (see above)
         """
         return clone(self)
 
+    @classmethod
+    def _get_init_signature(cls):
+        """Get init sigature of cls, for use in parameter inspection.
+
+        Returns
+        -------
+        list of inspect Parameter objects (including defaults)
+
+        Raises
+        ------
+        RuntimeError if cls has varargs in __init__
+        """
+        # fetch the constructor or the original constructor before
+        # deprecation wrapping if any
+        init = getattr(cls.__init__, "deprecated_original", cls.__init__)
+        if init is object.__init__:
+            # No explicit constructor to introspect
+            return []
+
+        # introspect the constructor arguments to find the model parameters
+        # to represent
+        init_signature = inspect.signature(init)
+
+        # Consider the constructor parameters excluding 'self'
+        parameters = [
+            p
+            for p in init_signature.parameters.values()
+            if p.name != "self" and p.kind != p.VAR_KEYWORD
+        ]
+        for p in parameters:
+            if p.kind == p.VAR_POSITIONAL:
+                raise RuntimeError(
+                    "scikit-learn compatible estimators should always "
+                    "specify their parameters in the signature"
+                    " of their __init__ (no varargs)."
+                    " %s with constructor %s doesn't "
+                    " follow this convention." % (cls, init_signature)
+                )
+        return parameters
+
+    @classmethod
+    def get_param_names(cls):
+        """Get parameter names for the object.
+
+        Returns
+        -------
+        param_names: list of str, alphabetically sorted list of parameter names of cls
+        """
+        parameters = cls._get_init_signature()
+        param_names = sorted([p.name for p in parameters])
+        return param_names
+
+    @classmethod
+    def get_param_defaults(cls):
+        """Get parameter defaults for the object.
+
+        Returns
+        -------
+        default_dict: dict with str keys
+            keys are all parameters of cls that have a default defined in __init__
+            values are the defaults, as defined in __init__
+        """
+        parameters = cls._get_init_signature()
+        default_dict = {
+            x.name: x.default for x in parameters if x.default != inspect._empty
+        }
+        return default_dict
+
+    def set_params(self, **params):
```

Comment:
Personally - I think one great thing to consider building in to the set_params function (since we are building our own) is a check for type!! I did a very sketchy version of this in my draft PR of the HMM estimator, but I think an even better way to do it would be to make use of the inspect signature library (ie can get annotations for parameters if type hints are provided).  Putting in a check like this can ensure that variables always remain the type intended.  Anyway - just an idea.  Seems like we are moving in the direction of using more type hints though - so could be a good idea?

Reply:
If at all, it should not happen in `set_params` but in `__init__`.
Whether it should happen in init or not is part of this design discussion https://github.com/alan-turing-institute/sktime/issues/2573, feel free to put your opinions there!